### PR TITLE
added testcases for hex and micro

### PR DESCRIPTION
**Pull request by Govintharaj**
I have written unittest for hexfile.py and micro.py file.
In that hexfile.py three test cases are failed, Because data() method in hexfile.py doesn't give all data from the given hex record line.
In hexfile.py  line number 104 there is logical error.
Error: for i in range(0, dlen, 2):
Expected: for i in range(0, (dlen*2), 2):

### CODE REVIEW

File: smashlib/test_hex.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import errno 
+import builtins
+import io
+from .hexfile import Hex,HexFile,HexError
+
+class SimpleTest(unittest.TestCase):
+	hex_data = Hex(b":03000000020008")
+	byte = io.BytesIO(b":03000000020008f3")
+	builtins.open = Mock(return_value=byte)
+	hex_read = HexFile(byte)
```
Comment (bravegnu) [inline smashlib/test_hex.py]: Move these into the test functions.

File: smashlib/test_hex.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	def test_count_error(self):
+		error = io.BytesIO(b":03000000")
+		error.name = Mock()
+		error.name.return_value = "test"
+		builtins.open = Mock(return_value = error)
+		self.assertRaises(HexError, HexFile(error).count_data_bytes)
+
+	def test_file_error(self):
+		error = io.BytesIO(b":03000001020008F3")
+		builtins.open = Mock(return_value = error)
```
Comment (bravegnu) [inline smashlib/test_hex.py]: Revert the mock once done.

File: smashlib/test_hex.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
+import unittest
+from unittest.mock import Mock
+import errno 
+import builtins
+import io
+from .hexfile import Hex,HexFile,HexError
```
Comment (bravegnu) [inline smashlib/test_hex.py]: Run pylint and check the coding style.

File: smashlib/test_hex.py [hunk 4]
```diff
@@ -1,101 +1,108 @@
+import io
 import unittest
 from unittest.mock import Mock
-import errno 
-import builtins
-import io
-from .hexfile import Hex,HexFile,HexError
+from unittest.mock import patch
```
Comment (Godson-Gnanaraj) [inline smashlib/test_hex.py]: you could have imported it in single line

File: smashlib/test_hex.py [hunk 5]
```diff
@@ -1,101 +1,108 @@
...
+    def test_hexfile_error(self):
+        self.assertRaises(HexError, Hex(":dfgh").data)
+
+    def test_addr_error(self):
+        self.assertRaises(HexError, Hex(":dfgh").addr)
+
+    def test_datalen_error(self):
+        self.assertRaises(HexError, Hex(":ggg").datalen)
+
+    def test_count_error(self):
```
Comment (Godson-Gnanaraj) [inline smashlib/test_hex.py]: setUp is not necessary for below tests

File: smashlib/test_micro.py [hunk 6]
```diff
@@ -1,123 +1,120 @@
+import io
 import unittest
 from unittest.mock import Mock
-from smashlib.hexfile import Hex, HexFile, HexError
-from smashlib.smash import Serial
+from unittest.mock import patch
```
Comment (Godson-Gnanaraj) [inline smashlib/test_micro.py]: can be imported in single line

File: smashlib/test_micro.py [hunk 7]
```diff
@@ -1,123 +1,120 @@
...
+
+    def test_dispdata(self):
+        hex_disp = HexDispData(0x20, 0x8)
+        expected_value = b":050000040020000800cf"
+        self.assertEqual(hex_disp.hex, expected_value)
+
+    def test_blank_check(self):
+        hex_blank = HexBlankCheck(0x20, 0x8)
+        expected_value = b":050000040020000801ce"
+        self.assertEqual(hex_blank.hex, expected_value)
```
Comment (Godson-Gnanaraj) [inline smashlib/test_micro.py]: can be tested separately.

File: smashlib/test_micro.py [hunk 8]
```diff
@@ -1,123 +1,120 @@
...
+        self.serial.wait_for.return_value = b"R"
+        self.assertRaises(IspProgError, self.micro_obj._send_cmd, b"U", None)
+
+    def test_send_cmd_progem_error_3(self):
+        self.serial.wait_for.return_value = b"P"
+        self.assertRaises(IspProgError, self.micro_obj._send_cmd, b"U", None)
+
+    def test_send_cmd_progem_error_2(self):
+        self.serial.wait_for.return_value = b"P"
+        self.assertRaises(IspProgError, self.micro_obj._send_cmd, b"U", None)
```
Comment (Godson-Gnanaraj) [inline smashlib/test_micro.py]: Can be tested using public methods

File: smashlib/test_micro.py [hunk 9]
```diff
@@ -1,123 +1,120 @@
...
+        self.assertRaisesRegex(ValueError, "data start address 0x54321 out of range",
+                               HexDispData, 0x54321, 0x4321)
+        self.assertRaisesRegex(ValueError, "data end address 0x43266 out of range",
+                               HexDispData, 0x5432, 0x43266)
+
+    def test_hex_dip_error(self):
+        self.assertRaisesRegex(ValueError, "check start address 0x54321 out of range",
+                               HexBlankCheck, 0x54321, 0x4321)
+        self.assertRaisesRegex(ValueError, "check end address 0x43266 out of range",
+                               HexBlankCheck, 0x5432, 0x43266)
```
Comment (Godson-Gnanaraj) [inline smashlib/test_micro.py]: can be tested separately.

File: smashlib/test_micro.py [hunk 10]
```diff
@@ -1,123 +1,120 @@
...
+    def test__update_cache_with_line_error(self):
+        self.assertRaises(ProtoError, self.micro_obj._update_cache_with_line, b'0x0000+0E')
+
+    def test__update_cache_with_line_addr_error(self):
+        self.assertRaises(ProtoError, self.micro_obj._update_cache_with_line, b'0xghgj=0E')
+
+    def test_getitem_error(self):
+        data = [b'0x0001=0E', b'']
+        self.micro_obj.serial.read_timeo = Mock(side_effect=data)
+        self.assertRaises(ProtoError, self.micro_obj._getitem, 0x0001)
```
Comment (Godson-Gnanaraj) [inline smashlib/test_micro.py]: test with public methods.

File: smashlib/test_micro.py [hunk 11]
```diff
@@ -1,123 +1,120 @@
+import io
```
Comment (Godson-Gnanaraj) [inline smashlib/test_micro.py]: Add missing testcases.

File: smashlib/test_micro.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+    def test_dispdata(self):
+        hex_disp = HexDispData(0x20, 0x8)
+        expected_value = b":050000040020000800cf"
+        self.assertEqual(hex_disp.hex, expected_value)
+
+    def test_blank_check(self):
+        hex_blank = HexBlankCheck(0x20, 0x8)
+        expected_value = b":050000040020000801ce"
+        self.assertEqual(hex_blank.hex, expected_value)
```
Comment (Godson-Gnanaraj) [inline smashlib/test_micro.py]: can be tested separately.

File: smashlib/test_micro.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
+import io
+import unittest
+from unittest.mock import Mock
+from unittest.mock import patch
```
Comment (Godson-Gnanaraj) [inline smashlib/test_micro.py]: can be imported in single line

File: smashlib/test_micro.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+        self.serial.wait_for.return_value = b"R"
+        self.assertRaises(IspProgError, self.micro_obj._send_cmd, b"U", None)
+
+    def test_send_cmd_progem_error_3(self):
+        self.serial.wait_for.return_value = b"P"
+        self.assertRaises(IspProgError, self.micro_obj._send_cmd, b"U", None)
+
+    def test_send_cmd_progem_error_2(self):
+        self.serial.wait_for.return_value = b"P"
+        self.assertRaises(IspProgError, self.micro_obj._send_cmd, b"U", None)
```
Comment (Godson-Gnanaraj) [inline smashlib/test_micro.py]: Can be tested using public methods

File: smashlib/test_micro.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+        self.assertRaisesRegex(ValueError, "data start address 0x54321 out of range",
+                               HexDispData, 0x54321, 0x4321)
+        self.assertRaisesRegex(ValueError, "data end address 0x43266 out of range",
+                               HexDispData, 0x5432, 0x43266)
+
+    def test_hex_dip_error(self):
+        self.assertRaisesRegex(ValueError, "check start address 0x54321 out of range",
+                               HexBlankCheck, 0x54321, 0x4321)
+        self.assertRaisesRegex(ValueError, "check end address 0x43266 out of range",
+                               HexBlankCheck, 0x5432, 0x43266)
```
Comment (Godson-Gnanaraj) [inline smashlib/test_micro.py]: can be tested separately.

File: smashlib/test_micro.py [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+    def test__update_cache_with_line_error(self):
+        self.assertRaises(ProtoError, self.micro_obj._update_cache_with_line, b'0x0000+0E')
+
+    def test__update_cache_with_line_addr_error(self):
+        self.assertRaises(ProtoError, self.micro_obj._update_cache_with_line, b'0xghgj=0E')
+
+    def test_getitem_error(self):
+        data = [b'0x0001=0E', b'']
+        self.micro_obj.serial.read_timeo = Mock(side_effect=data)
+        self.assertRaises(ProtoError, self.micro_obj._getitem, 0x0001)
```
Comment (Godson-Gnanaraj) [inline smashlib/test_micro.py]: can be tested using public methods.

File: smashlib/test_hex.py [hunk 17]
```diff
@@ -0,0 +1,108 @@
+import io
+import unittest
+from unittest.mock import Mock
+from unittest.mock import patch
```
Comment (Godson-Gnanaraj) [inline smashlib/test_hex.py]: you could have imported it in single line

File: smashlib/test_hex.py [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+    def test_hexfile_error(self):
+        self.assertRaises(HexError, Hex(":dfgh").data)
+
+    def test_addr_error(self):
+        self.assertRaises(HexError, Hex(":dfgh").addr)
+
+    def test_datalen_error(self):
+        self.assertRaises(HexError, Hex(":ggg").datalen)
+
+    def test_count_error(self):
```
Comment (Godson-Gnanaraj) [inline smashlib/test_hex.py]: "setUp" is not necessary for below tests. These can be tested in separate test class

File: smashlib/test_hex.py [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+from unittest.mock import Mock, patch
+from hexfile import Hex, HexFile, HexError
+
+class SimpleTest(unittest.TestCase):
+    def setUp(self):
+        self.hex_data = Hex(b":03000000020008")
+        string = io.BytesIO(b":03000000020008f3")
+        m_open = Mock(return_value=string)
+        with patch('builtins.open', m_open):
+            self.hex_read = HexFile('')
```
Comment (bravegnu) [inline smashlib/test_hex.py]: Why is this part of the setUp? This should be part of the test function.

File: smashlib/test_micro.py [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+                               HexBlankCheck, 0x54321, 0x4321)
+        self.assertRaisesRegex(ValueError, "check end address 0x43266 out of range",
+                               HexBlankCheck, 0x5432, 0x43266)
+
+class MockTestCase(unittest.TestCase):
+    def setUp(self):
+        self.serial = Mock()
+        self.micro_obj = Micro("P89V51RD2", 12, self.serial)
+
+    def data(self):
```
Comment (bravegnu) [inline smashlib/test_micro.py]: Can we have a better name?

File: smashlib/test_micro.py [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+        self.assertRaisesRegex(ValueError, "data end address 0x43266 out of range",
+                               HexDispData, 0x5432, 0x43266)
+
+    def test_hex_dip_error(self):
+        self.assertRaisesRegex(ValueError, "check start address 0x54321 out of range",
+                               HexBlankCheck, 0x54321, 0x4321)
+        self.assertRaisesRegex(ValueError, "check end address 0x43266 out of range",
+                               HexBlankCheck, 0x5432, 0x43266)
+
+class MockTestCase(unittest.TestCase):
```
Comment (bravegnu) [inline smashlib/test_micro.py]: Deserves a better name?

File: smashlib/test_hex.py [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+    def test_hexfile_error(self):
+        self.assertRaises(HexError, Hex(":dfgh").data)
+
+    def test_addr_error(self):
+        self.assertRaises(HexError, Hex(":dfgh").addr)
+
+    def test_datalen_error(self):
+        self.assertRaises(HexError, Hex(":ggg").datalen)
+
+class SecondTest(unittest.TestCase):
```
Comment (bravegnu) [inline smashlib/test_hex.py]: Deserves a better name?

File: smashlib/test_hex.py [hunk 23]
```diff
@@ -0,0 +1,117 @@
+import io
+import unittest
+from unittest.mock import Mock, patch
+from hexfile import Hex, HexFile, HexError
+
+class SimpleTest(unittest.TestCase):
```
Comment (bravegnu) [inline smashlib/test_hex.py]: Deserves a better name?

File: smashlib/test_hex.py [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+
+    def test_get_hex(self):
+        self.assertEqual(self.hex_data.get_hex(), b":03000000020008")
+
+    def test__bytes__(self):
+        self.assertEqual(self.hex_data.__bytes__(), b":03000000020008")
+
+    def test_block_from_addr(self):
+        with patch('builtins.open', self.m_open):
+            self.hex_read = HexFile('')
```
Comment (bravegnu) [inline smashlib/test_hex.py]: 'self' required?

File: smashlib/test_hex.py [hunk 25]
```diff
@@ -0,0 +1,125 @@
+import io
+import unittest
+from unittest.mock import Mock, patch
+from .hexfile import Hex, HexFile, HexError
+
+class HexTest(unittest.TestCase):
+
+    def setUp(self):
+        self.hex_data = Hex(b":03000000020008")
+        string = io.BytesIO(b":03000000020008f3")
```
Comment (bravegnu) [inline smashlib/test_hex.py]: Rename string to hex_file

File: smashlib/test_hex.py [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+    def test_data_bytes(self):
+        with patch('builtins.open', self.m_open):
+            self.hex_read = HexFile('')
+        self.assertEqual(self.hex_read.count_data_bytes(), 3)
+
+    def test_data_data(self):
+        data = self.hex_read.data_bytes()
+        data_list = []
+        for i in data:
+            data_list.append(i)
```
Comment (bravegnu) [inline smashlib/test_hex.py]: Will this work: list(data)?

File: smashlib/test_hex.py [hunk 27]
```diff
@@ -0,0 +1,125 @@
...
+    def test_get_hex(self):
+        self.assertEqual(self.hex_data.get_hex(), b":03000000020008")
+
+    def test__bytes__(self):
+        self.assertEqual(self.hex_data.__bytes__(), b":03000000020008")
+
+    def test_block_from_addr(self):
+        with patch('builtins.open', self.m_open):
+            self.hex_read = HexFile('')
+        self.assertEqual(self.hex_read.block_from_addr(1010, [(0000, 1111)]), 0)
```
Comment (bravegnu) [inline smashlib/test_hex.py]: Too much logic in a single line? Split this into multiple lines.

1. Arrange
2. Act
3. Assert

File: smashlib/test_hex.py [hunk 28]
```diff
@@ -0,0 +1,125 @@
...
+    def test_hexfile_error(self):
+        self.assertRaises(HexError, Hex(":dfgh").data)
+
+    def test_addr_error(self):
+        self.assertRaises(HexError, Hex(":dfgh").addr)
+
+    def test_datalen_error(self):
+        self.assertRaises(HexError, Hex(":ggg").datalen)
+
+class HexfileTest(unittest.TestCase):
```
Comment (bravegnu) [inline smashlib/test_hex.py]: HexFileTest

File: smashlib/test_micro.py [hunk 29]
```diff
@@ -0,0 +1,106 @@
...
+        with patch('builtins.open', self.data()):
+            self.assertRaises(IspChecksumError, self.micro_obj.prog_file, "")
+
+    def test_send_cmd_progem_error(self):
+        self.serial.wait_for.return_value = b"R"
+        string = io.BytesIO(b":03000000020008")
+        m_open = Mock(return_value=string)
+        with patch('builtins.open', m_open):
+            self.assertRaises(IspProgError, self.micro_obj.prog_file, "")   
+    def test_send_cmd_progem_error_3(self):
```
Comment (bravegnu) [inline smashlib/test_micro.py]: Spacing?

File: smashlib/test_micro.py [hunk 30]
```diff
@@ -0,0 +1,106 @@
...
+        self.assertEqual(self.micro_obj.serial.wait_for.called, True)
+        self.serial.wait_for.assert_called_with(b"U")
+
+    def test_sync_baudrate_error(self):
+        self.serial.wait_for.side_effect = IspTimeoutError('')
+        self.assertRaises(IspTimeoutError, self.micro_obj.sync_baudrate)
+
+    def test_program_file(self):
+        self.serial.wait_for.return_value = b"."
+        with patch('builtins.open', self.data()):
```
Comment (bravegnu) [inline smashlib/test_micro.py]: self.data()?

**Pull request merged**